pkg/controller/kubeconfig.go

```diff
@@ -5,61 +5,101 @@ import (
 	"time"
 
 	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/rand"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
 const (
 	secretTypeBootstrapToken v1.SecretType = "bootstrap.kubernetes.io/token"
+	tokenIDKey                             = "token-id"
+	tokenSecretKey                         = "token-secret"
+	tokenFormatter                         = "%s.%s"
+	tokenNameFormatter                     = "bootstrap-token-%s"
 )
 
+func (c *Controller) createBootstrapKubeconfig(name string) (*clientcmdapi.Config, error) {
+	token, err := c.createBootstrapToken(name)
+	if err != nil {
+		return nil, err
+	}
+
+	infoKubeconfig, err := c.kubeconfigProvider.GetKubeconfig()
+	if err != nil {
+		return nil, err
+	}
+
+	outConfig := infoKubeconfig.DeepCopy()
+
+	outConfig.AuthInfos = map[string]*clientcmdapi.AuthInfo{
+		"": {
+			Token: token,
+		},
+	}
+
+	return outConfig, nil
+}
+
 func (c *Controller) createBootstrapToken(name string) (string, error) {
+	existingSecret, err := c.getSecretIfExists(name)
+	if err != nil {
+		return "", err
+	}
+	if existingSecret != nil {
+		return c.updateSecretExpirationAndGetToken(existingSecret)
+	}
+
 	tokenID := rand.String(6)
 	tokenSecret := rand.String(16)
 
 	secret := v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Name: fmt.Sprintf(tokenNameFormatter, name),
 		},
 		Type: secretTypeBootstrapToken,
 		StringData: map[string]string{
 			"description":                    "bootstrap token for " + name,
-			"token-id":                       tokenID,
-			"token-secret":                   tokenSecret,
+			tokenIDKey:                       tokenID,
+			tokenSecretKey:                   tokenSecret,
 			"expiration":                     metav1.Now().Add(24 * time.Hour).Format(time.RFC3339),
 			"usage-bootstrap-authentication": "true",
 			"usage-bootstrap-signing":        "true",
 			"auth-extra-groups":              "system:bootstrappers:machine-controller:default-node-token",
 		},
 	}
 
-	_, err := c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Create(&secret)
+	_, err = c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Create(&secret)
 	if err != nil {
 		return "", err
 	}
 
-	return fmt.Sprintf("%s.%s", tokenID, tokenSecret), nil
+	return fmt.Sprintf(tokenFormatter, tokenID, tokenSecret), nil
 }
 
-func (c *Controller) createBootstrapKubeconfig(name string) (*clientcmdapi.Config, error) {
-	token, err := c.createBootstrapToken(name)
+func (c *Controller) updateSecretExpirationAndGetToken(secret *v1.Secret) (string, error) {
+	if secret.StringData == nil {
+		secret.StringData = map[string]string{}
+	}
+	secret.StringData["expiration"] = metav1.Now().Add(24 * time.Hour).Format(time.RFC3339)
```

Comment:
Maybe we should check if the token will expire within the next minutes and only then update it?
WDYT @p0lyn0mial 
It could help to reduce the load

Reply:
happy path doesn't have burden on performance it is one-off operation. 

I would leave it as it is, mainly because I don't want to deal with `Data["expiration"]` which is defined as `[]bytes`. I'd prefer to treat it as opaque value.

Unless we think this is important I can try to implement that.

Reply:
I was thinking about the PUT request a few lines down.